src/content/docs/en/reference/api-reference.mdx

```diff
@@ -861,301 +251,365 @@ A boolean representing whether or not the current page is prerendered.
+      return rewrite(new URL("../", Astro.url));
+    }
+    ```
+  </TabItem>
+</Tabs>
 
-See also: [`Astro.redirect()`](#astroredirect)
+Use a `Request` type for complete control of the `Request` sent to the server for the new path. The following example sends a request to render the parent page while also providing headers:
 
-### `context.rewrite()`
+<Tabs>
+  <TabItem label="Astro.rewrite()">
+    ```astro title="src/pages/blog/index.astro"
+    ---
+    return Astro.rewrite(new Request(new URL("../", Astro.url), {
+      headers: {
+        "x-custom-header": JSON.stringify(Astro.locals.someValue)
+      }
+    }))
+    ---
+    ```
+  </TabItem>
+  <TabItem label="context.rewrite()">
+    ```ts
+    import type { APIContext } from 'astro';
+
+    export function GET({ rewrite }: APIContext) {
+      return rewrite(new Request(new URL("../", Astro.url), {
+        headers: {
+          "x-custom-header": JSON.stringify(Astro.locals.someValue)
+        }
+      }));
+    }
+    ```
+  </TabItem>
+</Tabs>
+
+### `locals`
 
 <p>
 
-**Type:** `(rewritePayload: string | URL | Request) => Promise<Response>`<br />
-<Since v="4.13.0" />
+<Since v="2.4.0" />
 </p>
 
-Allows you to serve content from a different URL or path without redirecting the browser to a new page. 
+`locals` is an object used to store and access arbitrary information during the lifecycle of a request. `Astro.locals` is an object containing any values from the `context.locals` object from a middleware. Use this to access data returned by middleware in your `.astro` files.
 
-The method accepts either a string, a `URL`, or a `Request` for the location of the path.
-
-Use a string to provide an explicit path:
+Middleware functions can both read and write the values of `context.locals`:
 
-```ts
-import type { APIContext } from 'astro';
+```ts title="src/middleware.ts" "locals"
+import type { MiddlewareHandler } from 'astro';
 
-export function GET({ rewrite }: APIContext) {
-  return rewrite('/login');
+export const onRequest: MiddlewareHandler = ({ locals }, next) => {
+  if (!locals.title) {
+    locals.title = "Default Title";
+  }
+  return next();
 }
 ```
 
-Use a `URL` type when you need to construct the URL path for the rewrite. The following example renders a page's parent path by creating a new URL from the relative  `"../"` path:
+`Astro.locals` is read-only for Astro components and `context.locals` is read-only for API endpoints:
+
+<Tabs>
+  <TabItem label="Astro.locals">
+    ```astro title="src/pages/Orders.astro" "Astro.locals"
+    ---
+    const title = Astro.locals.welcomeTitle();
+    const orders = Array.from(Astro.locals.orders.entries());
+    ---
+    <h1>{title}</h1>
+    <ul>
+        {orders.map(order => {
+            return <li>{/* do something with each order */}</li>
+        })}
+    </ul>
+    ```
+  </TabItem>
+  <TabItem label="context.locals">
+    ```ts title="src/pages/hello.ts" "Astro.locals"
```

Comment:
nit, highlighting. This the context tabs, so `Astro.locals` does not work, I think this should be:
```suggestion
    ```ts title="src/pages/hello.ts" "locals"
```